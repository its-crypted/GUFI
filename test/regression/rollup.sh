#!/usr/bin/env bash
# This file is part of GUFI, which is part of MarFS, which is released
# under the BSD license.
#
#
# Copyright (c) 2017, Los Alamos National Security (LANS), LLC
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without modification,
# are permitted provided that the following conditions are met:
#
# 1. Redistributions of source code must retain the above copyright notice, this
# list of conditions and the following disclaimer.
#
# 2. Redistributions in binary form must reproduce the above copyright notice,
# this list of conditions and the following disclaimer in the documentation and/or
# other materials provided with the distribution.
#
# 3. Neither the name of the copyright holder nor the names of its contributors
# may be used to endorse or promote products derived from this software without
# specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
# IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
# INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
# LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
# OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
# ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
#
#
# From Los Alamos National Security, LLC:
# LA-CC-15-039
#
# Copyright (c) 2017, Los Alamos National Security, LLC All rights reserved.
# Copyright 2017. Los Alamos National Security, LLC. This software was produced
# under U.S. Government contract DE-AC52-06NA25396 for Los Alamos National
# Laboratory (LANL), which is operated by Los Alamos National Security, LLC for
# the U.S. Department of Energy. The U.S. Government has rights to use,
# reproduce, and distribute this software.  NEITHER THE GOVERNMENT NOR LOS
# ALAMOS NATIONAL SECURITY, LLC MAKES ANY WARRANTY, EXPRESS OR IMPLIED, OR
# ASSUMES ANY LIABILITY FOR THE USE OF THIS SOFTWARE.  If software is
# modified to produce derivative works, such modified software should be
# clearly marked, so as not to confuse it with the version available from
# LANL.
#
# THIS SOFTWARE IS PROVIDED BY LOS ALAMOS NATIONAL SECURITY, LLC AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
# THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED. IN NO EVENT SHALL LOS ALAMOS NATIONAL SECURITY, LLC OR
# CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
# EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT
# OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
# IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY
# OF SUCH DAMAGE.



set -e

ROOT="$(realpath ${BASH_SOURCE[0]})"
ROOT="$(dirname ${ROOT})"
ROOT="$(dirname ${ROOT})"
ROOT="$(dirname ${ROOT})"

GUFI_DIR2INDEX="${ROOT}/src/gufi_dir2index"
ROLLUP="${ROOT}/src/rollup"

TMP="tmp"
SRC="prefix"
INDEXROOT="${SRC}.gufi"
OUTPUT="rollup.out"

function cleanup() {
    rm -rf ${TMP} ${SRC} ${INDEXROOT}
}

trap cleanup EXIT

cleanup

(
# not rolled up:
#     1: ${TMP}
# rolled up:
#     16: (3 subdirs * [o+rx, ugo, ug, u]) + [o+rx, ugo, ug, u]
# files:
#     24: (6 files * [o+rx, ugo, ug, u])
mkdir ${TMP}                    # 0
mkdir -m 005 ${TMP}/o+rx        #  1
mkdir -m 007 ${TMP}/o+rx/dir1   #   1
touch ${TMP}/o+rx/dir1/file1
mkdir -m 005 ${TMP}/o+rx/dir2   #   1
touch ${TMP}/o+rx/dir2/file1
touch ${TMP}/o+rx/dir2/file2
mkdir -m 007 ${TMP}/o+rx/dir3   #   1
touch ${TMP}/o+rx/dir3/file1
touch ${TMP}/o+rx/dir3/file2
touch ${TMP}/o+rx/dir3/file3
mkdir -m 776 ${TMP}/ugo         #  4
mkdir -m 776 ${TMP}/ugo/dir1    #   1
touch ${TMP}/ugo/dir1/file1
mkdir -m 776 ${TMP}/ugo/dir2    #   1
touch ${TMP}/ugo/dir2/file1
touch ${TMP}/ugo/dir2/file2
mkdir -m 776 ${TMP}/ugo/dir3    #   1
touch ${TMP}/ugo/dir3/file1
touch ${TMP}/ugo/dir3/file2
touch ${TMP}/ugo/dir3/file3
mkdir -m 770 ${TMP}/ug          #  2
mkdir -m 770 ${TMP}/ug/dir1     #   1
touch ${TMP}/ug/dir1/file1
mkdir -m 773 ${TMP}/ug/dir2     #   1
touch ${TMP}/ug/dir2/file1
touch ${TMP}/ug/dir2/file2
mkdir -m 770 ${TMP}/ug/dir3     #   1
touch ${TMP}/ug/dir3/file1
touch ${TMP}/ug/dir3/file2
touch ${TMP}/ug/dir3/file3
mkdir -m 700 ${TMP}/u           #  3
mkdir -m 700 ${TMP}/u/dir1      #   1
touch ${TMP}/u/dir1/file1
mkdir -m 703 ${TMP}/u/dir2      #   1
touch ${TMP}/u/dir2/file1
touch ${TMP}/u/dir2/file2
mkdir -m 700 ${TMP}/u/dir3      #   1
touch ${TMP}/u/dir3/file1
touch ${TMP}/u/dir3/file2
touch ${TMP}/u/dir3/file3

# copy ${TMP} 3 times to different users
# 4 not rolled up: (3 * ${TMP}) + ${SRC}
# 48 rolled up: (3 * 16 dirs in ${TMP}
# 72 files: 3 * 24 files in ${TMP}
mkdir ${SRC}
cp -R ${TMP} ${SRC}/1001
chown -R 1001:1001 ${SRC}/1001
cp -R ${TMP} ${SRC}/1002
chown -R 1002:1002 ${SRC}/1002
cp -R ${TMP} ${SRC}/1003
chown -R 1003:1003 ${SRC}/1003
rm -r ${TMP}

${GUFI_DIR2INDEX} ${SRC} ${INDEXROOT}
${ROLLUP} -X ${INDEXROOT}
) 2>&1 | head -n -1 | tee "${OUTPUT}"

diff ${ROOT}/test/regression/rollup.expected "${OUTPUT}"
rm "${OUTPUT}"
