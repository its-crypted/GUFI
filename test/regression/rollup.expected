Ran 140 threads
Roots:
    prefix.gufi
Not Processed:               0

Not Rolled Up:               4
    Subdirectories:
        min:                 3
        max:                 4
        median:           4.00
        sum:                15
        average:          3.75
    Files/Links:
        min:                 0
        max:                 0
        median:           0.00
        sum:                 0
        average:          0.00
    Level:
        min:                 0
        max:                 1
        median:           1.00
        sum:                 3
        average:          0.75

Rolled Up:                  48
    Subdirectories:
        min:                 0
        max:                 3
        median:           0.00
        sum:                36
        average:          0.75
    Files/Links:
        min:                 0
        max:                 3
        median:           1.50
        sum:                72
        average:          1.50
    Level:
        min:                 2
        max:                 3
        median:           3.00
        sum:               132
        average:          2.75

    Successful:
        1:                   3
        2:                   3
        3:                   3
        4:                   3
        5:                  36
        Total:              48
    Failed:
        1:                   0
        2:                   0
        3:                   0
        4:                   0
        5:                   0
        Total:               0
Files/Links:    72
Directories:    52 (16 empty)
Total:          124
Remaining Dirs: 16 (30.77%)
$ gufi_query -d " " -S "SELECT path(name) from summary" "prefix.gufi" | wc -l
52

$ gufi_query -d " " -E "SELECT path(summary.name) || '/' || pentries.name from summary, pentries WHERE summary.inode == pentries.pinode" "prefix.gufi" | wc -l
72

$ gufi_query -d " " -S "SELECT path(name) from summary" -E "SELECT path(summary.name) || '/' || pentries.name from summary, pentries WHERE summary.inode == pentries.pinode" "prefix.gufi" | wc -l
124

$ gufi_find -type d | wc -l
52

$ gufi_find -type f | wc -l
72

$ gufi_find | wc -l
124

$ gufi_ls
1001
1002
1003

$ gufi_ls 1001
o+rx
u
ug
ugo

$ gufi_ls 1001/o+rx
dir1
dir2
dir3

$ gufi_ls 1001/o+rx/dir1
file1

$ gufi_ls 1001/o+rx/dir2
file1
file2

$ gufi_ls 1001/o+rx/dir3
file1
file2
file3

$ gufi_ls 1002
o+rx
u
ug
ugo

$ gufi_ls 1002/ugo/dir1
file1

$ gufi_ls 1002/ugo/dir2
file1
file2

$ gufi_ls 1002/ugo/dir3
file1
file2
file3

$ gufi_ls 1003
o+rx
u
ug
ugo

$ gufi_ls 1003/ug/dir1
file1

$ gufi_ls 1003/ug/dir2
file1
file2

$ gufi_ls 1003/ug/dir3
file1
file2
file3

# 1 less because gufi_ls does not list the input dir
$ gufi_ls -R | wc -l
123

$ gufi_stats    depth
0

$ gufi_stats -r depth
. 0
1001 1
1001/o+rx 2
1001/o+rx/dir1 3
1001/o+rx/dir2 3
1001/o+rx/dir3 3
1001/u 2
1001/u/dir1 3
1001/u/dir2 3
1001/u/dir3 3
1001/ug 2
1001/ug/dir1 3
1001/ug/dir2 3
1001/ug/dir3 3
1001/ugo 2
1001/ugo/dir1 3
1001/ugo/dir2 3
1001/ugo/dir3 3
1002 1
1002/o+rx 2
1002/o+rx/dir1 3
1002/o+rx/dir2 3
1002/o+rx/dir3 3
1002/u 2
1002/u/dir1 3
1002/u/dir2 3
1002/u/dir3 3
1002/ug 2
1002/ug/dir1 3
1002/ug/dir2 3
1002/ug/dir3 3
1002/ugo 2
1002/ugo/dir1 3
1002/ugo/dir2 3
1002/ugo/dir3 3
1003 1
1003/o+rx 2
1003/o+rx/dir1 3
1003/o+rx/dir2 3
1003/o+rx/dir3 3
1003/u 2
1003/u/dir1 3
1003/u/dir2 3
1003/u/dir3 3
1003/ug 2
1003/ug/dir1 3
1003/ug/dir2 3
1003/ug/dir3 3
1003/ugo 2
1003/ugo/dir1 3
1003/ugo/dir2 3
1003/ugo/dir3 3

$ gufi_stats    filecount
0

$ gufi_stats -r filecount
. 0
1001 0
1001/o+rx 0
1001/o+rx/dir1 1
1001/o+rx/dir2 2
1001/o+rx/dir3 3
1001/u 0
1001/u/dir1 1
1001/u/dir2 2
1001/u/dir3 3
1001/ug 0
1001/ug/dir1 1
1001/ug/dir2 2
1001/ug/dir3 3
1001/ugo 0
1001/ugo/dir1 1
1001/ugo/dir2 2
1001/ugo/dir3 3
1002 0
1002/o+rx 0
1002/o+rx/dir1 1
1002/o+rx/dir2 2
1002/o+rx/dir3 3
1002/u 0
1002/u/dir1 1
1002/u/dir2 2
1002/u/dir3 3
1002/ug 0
1002/ug/dir1 1
1002/ug/dir2 2
1002/ug/dir3 3
1002/ugo 0
1002/ugo/dir1 1
1002/ugo/dir2 2
1002/ugo/dir3 3
1003 0
1003/o+rx 0
1003/o+rx/dir1 1
1003/o+rx/dir2 2
1003/o+rx/dir3 3
1003/u 0
1003/u/dir1 1
1003/u/dir2 2
1003/u/dir3 3
1003/ug 0
1003/ug/dir1 1
1003/ug/dir2 2
1003/ug/dir3 3
1003/ugo 0
1003/ugo/dir1 1
1003/ugo/dir2 2
1003/ugo/dir3 3

$ gufi_stats    total-filecount
1001 24
1002 24
1003 24

$ gufi_stats -c total-filecount
72

$ gufi_stats    dircount
3

$ gufi_stats -r dircount
. 3
1001 4
1001/o+rx 3
1001/o+rx/dir1 0
1001/o+rx/dir2 0
1001/o+rx/dir3 0
1001/u 3
1001/u/dir1 0
1001/u/dir2 0
1001/u/dir3 0
1001/ug 3
1001/ug/dir1 0
1001/ug/dir2 0
1001/ug/dir3 0
1001/ugo 3
1001/ugo/dir1 0
1001/ugo/dir2 0
1001/ugo/dir3 0
1002 4
1002/o+rx 3
1002/o+rx/dir1 0
1002/o+rx/dir2 0
1002/o+rx/dir3 0
1002/u 3
1002/u/dir1 0
1002/u/dir2 0
1002/u/dir3 0
1002/ug 3
1002/ug/dir1 0
1002/ug/dir2 0
1002/ug/dir3 0
1002/ugo 3
1002/ugo/dir1 0
1002/ugo/dir2 0
1002/ugo/dir3 0
1003 4
1003/o+rx 3
1003/o+rx/dir1 0
1003/o+rx/dir2 0
1003/o+rx/dir3 0
1003/u 3
1003/u/dir1 0
1003/u/dir2 0
1003/u/dir3 0
1003/ug 3
1003/ug/dir1 0
1003/ug/dir2 0
1003/ug/dir3 0
1003/ugo 3
1003/ugo/dir1 0
1003/ugo/dir2 0
1003/ugo/dir3 0

$ gufi_stats    total-dircount
1001 17
1002 17
1003 17

$ gufi_stats -c total-dircount
51

$ gufi_stats    files-per-level
1001 3 24
1002 3 24
1003 3 24

$ gufi_stats -c files-per-level
3 72

$ gufi_stats    dirs-per-level
root 0 1
1001 1 1
1002 1 1
1003 1 1
1001 2 4
1002 2 4
1003 2 4
1001 3 12
1002 3 12
1003 3 12

$ gufi_stats -c dirs-per-level
0 1
1 3
2 12
3 36

